File: implementations/flutter/design_patterns_flutter/lib/patterns/behavioral/visitor/visitor_csv.dart
import 'visitor.dart';
import 'element_book.dart';
import 'element_electronics.dart';
import 'element_grocery.dart';

class CsvExportVisitor implements Visitor {
  final List<String> rows = ['type,name,qty,unitPrice,extra'];

  @override
  void visitBook(Book e) {
    rows.add('book,"${e.title.replaceAll('"', '""')}",${e.qty},${e.unitPrice},imported=${e.isImported}');
  }

  @override
  void visitElectronics(Electronics e) {
    rows.add('electronics,"${e.name.replaceAll('"', '""')}",${e.qty},${e.unitPrice},fragile=${e.fragile}');
  }

  @override
  void visitGrocery(Grocery e) {
    rows.add('grocery,"${e.name.replaceAll('"', '""')}",${e.qty},${e.unitPrice},perishable=${e.perishable}');
  }

  @override
  String toString() => rows.join('\n');
}

File: implementations/react-native/design-patterns-rn/src/patterns/behavioral/visitor/visitor-csv.ts
import type { Book } from './element-book';
import type { Electronics } from './element-electronics';
import type { Grocery } from './element-grocery';
import type { Visitor } from './visitor';

export class CsvExportVisitor implements Visitor {
  rows: string[] = ['type,name,qty,unitPrice,extra'];
  visitBook(e: Book) {
    this.rows.push(`book,${csv(e.title)},${e.qty},${e.unitPrice},imported=${e.isImported}`);
  }
  visitElectronics(e: Electronics) {
    this.rows.push(`electronics,${csv(e.name)},${e.qty},${e.unitPrice},fragile=${e.fragile}`);
  }
  visitGrocery(e: Grocery) {
    this.rows.push(`grocery,${csv(e.name)},${e.qty},${e.unitPrice},perishable=${e.perishable}`);
  }
  toString() { return this.rows.join('\n'); }
}
const csv = (s: string) => `"${s.replace(/"/g, '""') }"`;
